runtime/lua/vim/lsp/buf.lua

```diff
@@ -116,145 +116,85 @@ end
+--- vim.api.nvim_command[[autocmd BufWritePre <buffer> lua vim.lsp.buf.formatting({save = true})]]
 --- </pre>
 ---
----@param options (optional, table) `FormattingOptions` entries
----@param timeout_ms (optional, number) Request timeout
----@param order (optional, table) List of client names. Formatting is requested from clients
----in the following order: first all clients that are not in the `order` list, then
----the remaining clients in the order as they occur in the `order` list.
-function M.formatting_seq_sync(options, timeout_ms, order)
-  local clients = vim.tbl_values(vim.lsp.buf_get_clients());
-  local bufnr = vim.api.nvim_get_current_buf()
-
-  -- sort the clients according to `order`
-  for _, client_name in pairs(order or {}) do
-    -- if the client exists, move to the end of the list
-    for i, client in pairs(clients) do
-      if client.name == client_name then
-        table.insert(clients, table.remove(clients, i))
-        break
-      end
-    end
-  end
-
-  -- loop through the clients and make synchronous formatting requests
-  for _, client in pairs(clients) do
-    if client.resolved_capabilities.document_formatting then
-      local params = util.make_formatting_params(options)
-      local result, err = client.request_sync("textDocument/formatting", params, timeout_ms, vim.api.nvim_get_current_buf())
-      if result and result.result then
-        util.apply_text_edits(result.result, bufnr)
-      elseif err then
-        vim.notify(string.format("vim.lsp.buf.formatting_seq_sync: (%s) %s", client.name, err), vim.log.levels.WARN)
+---@param opts|nil
+---        * formatting_options table with valid `FormattingOptions` entries
+---        * client_id number restrict formatting to client_id
+---        * timeout_ms number Request timeout
+---        * save bool save buffer after formatting
+---@see https://microsoft.github.io/language-server-protocol/specification#textDocument_formatting
+function M.formatting(opts)
+  opts = opts or {}
+  local client = vim.lsp.get_client_by_id(opts.client_id) or select_client("textDocument/formatting")
+  if client == nil then return end
+
+  local request_buf = vim.api.nvim_get_current_buf()
+  vim.api.nvim_buf_set_option(request_buf, 'modifiable', false)
+  vim.notify(string.format("Locking buffer %d for formatting.", request_buf))
+
+  local params = util.make_formatting_params(opts.formatting_options)
+  local success, request_id = client.request("textDocument/formatting", params, function(_, result, ctx, _)
+      vim.api.nvim_buf_set_option(request_buf, 'modifiable', true)
+      if not result then return end
+      util.apply_text_edits(result, ctx.bufnr)
+      if opts.save then
+        vim.api.nvim_buf_call(request_buf, function() vim.api.nvim_command("noautocmd write!") end)
       end
-    end
+    end, request_buf)
+  if success then
+    client.requests[request_id].type = "blocking"
```

Comment:
This is kinda hacky, but the idea is that there are now "blocking" PRs for which we will wait 2s before closing neovim

Reply:
Did you try to see what kind of impact it would have if we just generally wait for all pending requests to complete before exiting?